File: stage.cps/Trainer.cs
using System.Collections.Generic;
using System.Linq;
using stage.domain;

namespace stage.cps {
    public static class Trainer {
        public static double CalculatePassiveCooling(List<Snapshot> snapshots) {
            List<double> deltaValues = new List<double>();
            for(int i = 0; i < snapshots.Count; i++) {

                Snapshot snapshot = snapshots[i];
                if (snapshot.HeatingPercentage == 0 && snapshot.CoolingPercentage == 0) {
                    double deltaValue = snapshots[i+1].Value - snapshot.Value;
                    deltaValues.Add(deltaValue);
                }
            }

            return deltaValues.Count == 0 ? 0 : (deltaValues.Sum() / deltaValues.Count);
        }

        public static double CalculateActiveHeating(List<Snapshot> snapshots) {
            List<double> deltaValues = new List<double>();
            for(int i = 0; i < snapshots.Count; i++) {

                Snapshot snapshot = snapshots[i];
                if (snapshot.HeatingPercentage > 0 && snapshot.CoolingPercentage == 0) {
                    double deltaValue = snapshots[i+1].Value - snapshot.Value;
                    deltaValues.Add(deltaValue);
                }
            }

            return deltaValues.Count == 0 ? 0 : (deltaValues.Sum() / deltaValues.Count);
        }

        public static double CalculateActiveCooling(List<Snapshot> snapshots) {
            List<double> deltaValues = new List<double>();
            for(int i = 0; i < snapshots.Count; i++) {

                Snapshot snapshot = snapshots[i];
                if (snapshot.HeatingPercentage == 0 && snapshot.CoolingPercentage > 0) {
                    double deltaValue = snapshots[i+1].Value - snapshot.Value;
                    deltaValues.Add(deltaValue);
                }
            }

            return deltaValues.Count == 0 ? 0 : (deltaValues.Sum() / deltaValues.Count);
        }
    }
}

File: src/cps/trainer.ts
import { Snapshot } from '../models/snapshot.model';

export class Trainer {
    public static calculatePassiveCooling(snapshots: Snapshot[]): number {
        const deltaValues: Array<number> = [];
        for(let i = 0; i < snapshots.length-1; i++) {

            const snapshot = snapshots[i];
            if (snapshot.heatingPercentage == 0 && snapshot.coolingPercentage == 0) {
                const deltaValue = snapshots[i+1].value - snapshot.value;
                deltaValues.push(deltaValue);
            }
        }

        return deltaValues.length == 0 ? 0 : (deltaValues.reduce((a, b) => a + b, 0) / deltaValues.length);
    }

    public static calculateActiveHeating(snapshots: Snapshot[]): number {
        const deltaValues: Array<number> = [];
        for(let i = 0; i < snapshots.length-1; i++) {

            const snapshot = snapshots[i];
            if (snapshot.heatingPercentage > 0 && snapshot.coolingPercentage == 0) {
                const deltaValue = snapshots[i+1].value - snapshot.value;
                deltaValues.push(deltaValue);
            }
        }

        return deltaValues.length == 0 ? 0 : (deltaValues.reduce((a, b) => a + b, 0) / deltaValues.length);
    }

    public static calculateActiveCooling(snapshots: Snapshot[]): number {
        const deltaValues: Array<number> = [];
        for(let i = 0; i < snapshots.length-1; i++) {

            const snapshot = snapshots[i];
            if (snapshot.heatingPercentage == 0 && snapshot.coolingPercentage > 0) {
                const deltaValue = snapshots[i+1].value - snapshot.value;
                deltaValues.push(deltaValue);
            }
        }

        return deltaValues.length == 0 ? 0 : (deltaValues.reduce((a, b) => a + b, 0) / deltaValues.length);
    }
}